src/ipa_pc/mod.rs

```diff
@@ -1204,146 +1223,214 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
 
         // h(X)
-        let h_poly_time = start_timer!(|| "Compute batching polynomial");
-        let mut batch_polynomial = Polynomial::zero();
+        let mut h_polynomial = Polynomial::zero();
+
+        // z(X)
+        let mut z_polynomial = Polynomial::from_coefficients_slice(
+            &[G::ScalarField::one()]
+        );
 
         for (label, (_point_label, point)) in query_set.iter() {
             let labeled_polynomial =
-                poly_map.get(label).ok_or(Error::MissingPolynomial {
+                *poly_map.get(label).ok_or(Error::MissingPolynomial {
                     label: label.to_string(),
                 })?;
 
             if labeled_polynomial.hiding_bound().is_some() {
                 has_hiding = true;
             }
 
-            points.push(point);
-
             // y_i
-            let evaluated_y = labeled_polynomial.polynomial().evaluate(*point);
+            let y_i = labeled_polynomial.polynomial().evaluate(*point);
+
+            // (X - x_i)
+            let x_polynomial = Polynomial::from_coefficients_slice(&[
+                (G::ScalarField::zero() - point),
+                G::ScalarField::one()
+            ]);
+
+            // z(X) = PROD (X - x_i)
+            z_polynomial = &z_polynomial * &x_polynomial;
 
             // (p_i(X) - y_i) / (X - x_i)
             let polynomial =
-                &(labeled_polynomial.polynomial() - &Polynomial::from_coefficients_vec(vec![evaluated_y]))
+                &(labeled_polynomial.polynomial() - &Polynomial::from_coefficients_slice(&[y_i]))
                     /
-                    &Polynomial::from_coefficients_vec(vec![
-                        (G::ScalarField::zero() - point),
-                        G::ScalarField::one()
-                    ]);
+                    &x_polynomial;
 
             // h(X) = SUM( lambda^i * ((p_i(X) - y_i) / (X - x_i)) )
-            batch_polynomial += (cur_challenge, &polynomial);
+            h_polynomial += (cur_challenge, &polynomial);
 
             // lambda^i
             cur_challenge = cur_challenge * &lambda;
         }
+
         end_timer!(h_poly_time);
 
         let key_len = ck.comm_key.len();
-        let p_len = batch_polynomial.coeffs.len();
-        let segments_count = std::cmp::max(1, p_len / key_len + if p_len % key_len != 0 { 1 } else { 0 });
-
-        let batch_randomness = if has_hiding {
-            if rng.is_none() {
-                Err(Error::Other("Rng not set".to_owned()))?
-            }
-            Randomness::rand(segments_count, false, rng.as_mut().unwrap())
-        } else {
-            Randomness::empty(segments_count)
-        };
 
-        // Commitment of the h(X) polynomial
-        let commit_time = start_timer!(|| format!("Commit to batch polynomial of degree {}", batch_polynomial.degree()));
-        let batch_commitment: Vec<G>;
+        let h_polynomial_len = h_polynomial.coeffs.len();
+        let lc_polynomial_len = h_polynomial_len + 1;
```

Comment:
This is an overestimate as it does not take into account the possibility that leading terms may cancel out. 
However, we won't  need it as we do not commit the polynomial of the LC (see my comment below).

Reply:
Done